controller/controller.go

```diff
@@ -0,0 +1,70 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+const DefaultPollingInterval = 10 * time.Second
+
+// Controller is a generic controller implementation.
+// It periodically polls for Resources and reconciles
+// them by calling Resource's Upsert or Delete function accordingly.
+type Controller struct {
+	// Resources lists resources for Controller to reconcile.
+	Resources ResourceLister
+	// PollingInterval is an interval that Controller will use to reconcile all Resources.
+	PollingInterval time.Duration
+	// Log is the logger used by the Controller.
+	Log hclog.Logger
+
+	// resourceState is the internal Controller's state. The Controller
+	// will periodically update the state to reflect the state of the Resources.
+	resourceState map[string]struct{}
```

Comment:
This is a map of the ids of the resources that we currently know to exist and that have successfully been upserted once right? Should we state that more precisely in the comment or are you leaving this a bit more vague as "internal state" because it might be used for something else later?

Reply:
In addition to the comment, we could add a type def to clarify this too, e.g.

```go
type ResourceId string

type Resource interface {
	ID() (ResourceId, error)
        ...
}

type Controller struct {
        resourceState map[ResourceId]struct{}
        ...
}
```

Totally optional.

Reply:
>This is a map of the ids of the resources that we currently know to exist and that have successfully been upserted once right

Right.

>In addition to the comment, we could add a type def to clarify this too, e.g.

Yeah, I like that idea!